### PR TITLE
#2366 -  ECert feedback file part 3

### PR DESCRIPTION
**Pull request by dheepak-aot**
# ECert feedback file part 3

## Start saving the integration file name that imported the records

- [x] DB migration to add the column `feedback_file_name`
- [x] Updated the e-cert feedback file handler to start saving the name of the integration file which imported the feedback errors.
- [x] Updated E2E test factory to save a dummy file name.

## E2E Tests

- [x] Created E2E tests for `ecert-part-time-feedback-integration.scheduler` and `ecert-full-time-feedback-integration.scheduler`

**Part Time**
image

**Full Time**
image

## E2E Process Summary error

- [x] The process summary framework has been updated to avoid logging the error stack trace in the event of custom named error considering that stack trace is not essential for a custom named error and also it introduces added complexity to parse and assert custom error messages in E2E tests.

image

## E2E Utility 

- [x] Updated the E2E Utility `getStructuredRecords()` to use the delimiter in SFTP service to be in sync with the process.

### CODE REVIEW

File: sources/packages/backend/libs/services/src/constants/error-code.constants.ts [hunk 1]
```diff
@@ -32,3 +32,4 @@ export const UNEXPECTED_ERROR_DOWNLOADING_FILE =
 export const ECE_DISBURSEMENT_DATA_NOT_VALID =
   "ECE_DISBURSEMENT_DATA_NOT_VALID";
 export const FILE_PARSING_ERROR = "FILE_PARSING_ERROR";
+export const DOCUMENT_NUMBER_NOT_FOUND = "DOCUMENT_NUMBER_NOT_FOUND";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/constants/error-code.constants.ts:35]: Is this not more close to the integrations lib? Is there a reason to be created into the services lib?

File: sources/packages/backend/libs/utilities/src/logger/process-summary.ts [hunk 2]
```diff
@@ -104,7 +105,17 @@ export class ProcessSummary {
     error: unknown = null,
     scope: LogScopes = LogScopes.Combined,
   ): void {
-    const errorDescription = error ? parseJSONError(error) : "";
+    let errorDescription = "";
+    // Considering that logging the stack trace is not useful for custom named errors
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/utilities/src/logger/process-summary.ts:109]: Can the message be adjusted as below?
```
// Considering that logging the stack trace is not useful for custom-named errors and only the message is sufficient,
// the custom-named errors are identified and the custom error message is added to the process summary.
// For any other runtime error, logging the stack trace is considered essential.
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+    it("Should log 'Invalid record type' error when the record type of header is invalid.", async () => {
+      // Arrange
+      mockDownloadFiles(
+        sftpClientMock,
+        [FEEDBACK_ERROR_FILE_SINGLE_RECORD],
+        (fileContent: string) => {
+          const file = getStructuredRecords(fileContent);
+          // Force the header to be wrong.
+          file.header = file.header.replace("100222  NEW", "500222  NEW");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts:75]: `replace` will change only the first match, hence the search string can be simplified as below.
```ts
file.header = file.header.replace("100", "500");
```
or
```ts
file.header = file.header.replace("1", "5");
```

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+
+    it("Should log 'Invalid record type' error when the record type of footer is invalid.", async () => {
+      // Arrange
+      mockDownloadFiles(
+        sftpClientMock,
+        [FEEDBACK_ERROR_FILE_SINGLE_RECORD],
+        (fileContent: string) => {
+          const file = getStructuredRecords(fileContent);
+          // Force the SIN hash total footer to be wrong.
+          file.footer = file.footer.replace("999 NEW", "777 NEW");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts:111]: Same from https://github.com/bcgov/SIMS/pull/3269/files#r1599041619

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    });
+
+    it("Should log 'Invalid record type' error when the record type of footer is invalid.", async () => {
+      // Arrange
+      mockDownloadFiles(
+        sftpClientMock,
+        [FEEDBACK_ERROR_FILE_SINGLE_RECORD],
+        (fileContent: string) => {
+          const file = getStructuredRecords(fileContent);
+          // Force the SIN hash total footer to be wrong.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts:110]: Is the comment meaningful for this method?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+          },
+        });
+      expect(disbursementFeedbackErrors.length).toBe(1);
+      const [importedError] = disbursementFeedbackErrors;
+      expect(importedError).toEqual({
+        id: expect.any(Number),
+        disbursementSchedule: { id: disbursementSchedule.id },
+        eCertFeedbackError: {
+          id: expect.any(Number),
+          errorCode: SHARED_ERROR_CODE,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-feedback-integration.scheduler.e2e-spec.ts:301]: Does it make sense to test the offering intensity also?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-part-time-feedback-integration.scheduler.e2e-spec.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-part-time-feedback-integration.scheduler.e2e-spec.ts:1]: Please take a look at the comments from the full-time E2E.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 7]
```diff
@@ -337,28 +338,42 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
         eCertFeedbackResponseRecords,
         eCertFeedbackErrorCodeMap,
       );
+      if (unknownErrorCodesMessage) {
+        // Abort the file processing and return after logging the unknown error codes.
+        processSummary.error(unknownErrorCodesMessage);
+        return;
+      }
+      // Get the file name from the file path.
+      const feedbackFileName = filePath.replace(
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:365]: you meant `path.basename()`, if so sure I can use it to get the file name.

File: sources/packages/backend/libs/services/src/constants/error-code.constants.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/constants/error-code.constants.ts:35]: Reason being we don't have error-code.constants at integration level. I can create one.

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/constants/error-code.constants.ts]: With the provided explanation there was no need to.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍

**Pull request merged**